### DISCUSSION TITLE
Help with findUnique returning null with Apollo graphql with BigInt id fields

### DISCUSSION THREAD

**Discussion by Raees678** [prisma/prisma#17824] [Q&A]
I am having issues with findUnique returning null, when I have my ID type as a BigInt. I have a Prisma schema like so:

```schema.prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "sqlite"
  url      = env("DATABASE_URL")
}

model User {
  id      BigInt    @id @default(autoincrement())
  email   String    @unique
  name    String?
  posts   Post[]
}

model Post {
  id        Int       @id @default(autoincrement())
  title     String
  content   String?
  published Boolean   @default(false)
  author    User?     @relation(fields: [authorId], references: [id])
  authorId  BigInt?
}
```

This is pretty similar to the starter example. The only difference is the ID is a BigInt type, since my DB requires a BigInt field for the ID. I then created an Apollo GraphQL server in an `index.ts` file like so, defining a BigInt scalar type in it, and adding it to the Apollo server. I have a single query `users` that returns a user and then uses Prisma's fluent API to return the nested posts.

```index.ts
import { ApolloServer } from "@apollo/server";
import gql from "graphql-tag";
import { PrismaClient } from "@prisma/client";
import express from "express";
import http from "http";
import cors from "cors";
import bodyParser from "body-parser";
import cookieParser from "cookie-parser";
import { expressMiddleware } from "@apollo/server/express4";
import { GraphQLScalarType, Kind } from "graphql";

const BigIntScalar = new GraphQLScalarType({
  name: "BigInt",
  description: "BigInt custom scalar type",
  parseValue: (value: unknown): bigint => {
    if (
      typeof value === "number" ||
      typeof value === "string" ||
      typeof value === "bigint" ||
      typeof value === "boolean"
    ) {
      return BigInt(value); // Convert incoming integer to BigInt
    }
    throw new Error("GraphQL BigInt Scalar parser expected a `number` or `string` or `bigint` or `boolean`");
  },
  // value sent to the client
  serialize: (value: unknown): unknown => {
    return parseInt(value as string);
  },
  // ast value is always in string format
  parseLiteral(ast): unknown {
    if (ast.kind === Kind.INT) {
      return ast.value;
    }
    return null;
  },
});

const typeDefs = gql`
  scalar BigInt

  type User {
    id: BigInt!
    email: String!
    name: String!
    posts: [Post!]!
  }
  type Post {
    id: BigInt!
    title: String!
    content: String
    published: Boolean!
    author: User
  }
  type Query {
    users: [User!]!
  }
`;

const client = new PrismaClient({
  log: ["query"],
});

const resolvers = {
  BigInt: BigIntScalar,
  Query: {
    users: async (_, __, contextValue) => {
      return contextValue.client.user.findMany({
        where: {
          id: {
            in: [1, 2],
          },
        },
      });
    },
  },
  User: {
    posts: async (user, _, contextValue) => {
      const posts = contextValue.client.user
        .findUnique({
          where: {
            id: user.id,
          },
        })
        .posts();

      return posts;
    },
  },
};

async function main() {
  const app = express();
  const httpServer = http.createServer(app);

  const server = new ApolloServer({
    typeDefs,
    resolvers,
    introspection: true,
  });

  await server.start();

  app.use(
    "/",
    cors<cors.CorsRequest>({
      credentials: true,
    }),
    bodyParser.json(),
    cookieParser(),
    expressMiddleware(server, {
      context: async () => {
        return {
          client,
        };
      },
    })
  );

  await new Promise<void>((resolve) => httpServer.listen({ port: 4001 }, resolve));
  console.log(`🚀 Server ready at http://localhost:4001`);
}

main();
```
I have seeded the DB with 2 users and created 2 posts for them. When I run a query like this:

```.graphql
query {
  users {
    name
    posts {
      title
    }
  }
}
```

I would expect to get 2 users, and their nested posts. However, I receive an error: `Cannot return null for non-nullable field User.posts`. 

When I change the user resolver to only return data for a single user like this:

```index.ts
    users: async (_, __, contextValue) => {
      return contextValue.client.user.findMany({
        where: {
          id: {
            in: [1],
          },
        },
      });
    },
```
or like this:
```index.ts
    users: async (_, __, contextValue) => {
      return contextValue.client.user.findMany({
        where: {
          id: {
            in: [2],
          },
        },
      });
    },
```
I do see the correct output with users and their posts being resolved.

Also if I change the type of the ID in the DB for the user from BigInt to Int, and also change the foreign keys in the schema to Int, then querying both users' posts resolves correctly.

- How would I go about diagnosing/fixing the issue here? 
- Is there something I am doing wrong? 
- Why would the BigInt type cause an issue when requesting 2 users and their posts but not the Int type?
- Why does BigInt not have an issue returning a single users' post but it returns null when requesting for more than 1 user?
- Is this an issue with Apollo or with Prisma?

**Comment by nurul3101** [comment]
Hey @Raees678 👋 

Can you try changing your User resolver to not use the fluent API and instead use `include` and check if you get the same behaviour?

```typescript
const posts = contextValue.client.user
        .findUnique({
          where: {
            id: user.id,
          },
          include:{
            posts: true
          }
        });
```

**Reply by Raees678** [reply]
I just tried this and yes, I get the same behavior. What could that mean?

**Reply by Raees678** [reply]
In fact if I `await` the value of the promise returned by the query and try to log it in the resolver itself I see that the `posts` variable resolves to `null`. This happens if I use the fluent api or the above method.

However if I query for only a single user, or I change the type from BigInt to Int and query for both users, the value of post resolves correctly to a post after awaiting the object in the resolver.

**Reply by nurul3101** [reply]
I just realised that you are using an SQLite database. Can you just momentarily replace it with Postgres database and see if this gets resolved. If it resolves, then we would be able to verify that the issue is with BigInt columns in a SQLite database.

**Reply by Raees678** [reply]
I tried the above with both a MySQL and a PostgreSQL database. With a MySQL database the underlying type in the DB was `bigint`, with Postgres the underlying type generated by Prisma migrate was `int8`. In both cases I faced the same error.

**Reply by Raees678** [reply]
After looking into this more it seems like:

```.typescript
contextValue.client.user
        .findUnique({
          where: {
            id: user.id,
          }
```

just this bit of code returns `null` when awaiting it, instead of returning an actual user. But it only happens within the confines of the resolver, and if the resolver is running for more than 1 user. 

If I try to run it outside the resolver just globally for eg. it works fine and returns an actual user.

**Reply by Raees678** [reply]
And when I log the actual queries it seems to be running them like a dataloader would as I expect (with the user query running IN 1,2) and the post query running only once:

```.ts
{
  timestamp: 2023-02-09T17:45:56.190Z,
  query: 'SELECT `hello-prisma`.`User`.`id`, `hello-prisma`.`User`.`email`, `hello-prisma`.`User`.`name` FROM `hello-prisma`.`User` WHERE `hello-prisma`.`User`.`id` IN (?,?)',
  params: '[1,2]',
  duration: 0,
  target: 'quaint::connector::metrics'
}
{
  timestamp: 2023-02-09T17:45:56.193Z,
  query: 'SELECT `hello-prisma`.`User`.`id` FROM `hello-prisma`.`User` WHERE `hello-prisma`.`User`.`id` IN (?,?)',
  params: '[1,2]',
  duration: 0,
  target: 'quaint::connector::metrics'
}
{
  timestamp: 2023-02-09T17:45:56.193Z,
  query: 'SELECT `hello-prisma`.`Post`.`id`, `hello-prisma`.`Post`.`title`, `hello-prisma`.`Post`.`content`, `hello-prisma`.`Post`.`published`, `hello-prisma`.`Post`.`authorId` FROM `hello-prisma`.`Post` WHERE `hello-prisma`.`Post`.`authorId` IN (?,?)',
  params: '[1,2]',
  duration: 0,
  target: 'quaint::connector::metrics'
}
```

The logs look the same if I use either BigInt or Int, but if I use Int awaiting the user within the resolver yields an actual User, but when using a BigInt it yields null.

**Reply by nurul3101** [reply]
That's strange, I am not entirely sure why the query behaves differently when invoked in the context of a resolver and when invoked globally. Are you perhaps using the latest version of your dependencies?

**Reply by Raees678** [reply]
Yup, using Prisma 4.10.1., and the latest version of apollo. I feel like the issue is arising from batched `findUnique` queries in prisma, and not something to do with apollo. 

This is because I can issue findUnique queries on `Int` or `BigInt` types in apollo and everything will run fine as long as Prisma's data type is not `BigInt`. If the `schama.prisma` defines it as an `Int`, the findUnique batched queries run fine, regardless if Apollo's serializer serializes it to an Int or a custom `BigInt` scalar in the response. However, if the `schema.prisma` defines it as a `BigInt` then the queries fail. So the issues must be arising from prisma's handling of the data being returned not any problem with apollo unable to serialize responses properly.

This could also explain why it works on a global level but not in the resolver. Since the prisma dataloader runs the `findUnique` queries in batches when in the resolver but just runs it normally globally.

**Reply by nurul3101** [reply]
Hmm, that's interesting. I'll try to reproduce this internally and can let our engineering team know about this behaviour.

**Reply by kwesterfeld2** [reply]
I am facing this exact issue

**Comment by edlundin** [comment]
Hi!

I ran into something similar after updating Prisma to 4.10.1 (I am also using PostgreSQL). When using `findUnique()`, it always returns `null`. However, I cannot reproduce the issue with `findUniqueOrThrow()`. 

The following snippet's, in version 4.10.1, `findUnique()` always returns `null`:

```typescript
try {
	const MODE = await this.prisma.mode.findUnique({ where: { id } }); //Always returns null.
	//const MODE = await this.prisma.mode.findUniqueOrThrow({ where: { id } }); //Would return the DB's record.

	if (!MODE) {
		throw new Error(`(ID: ${id}): mode not found`));
	}

	console.log(JSON.stringify(MODE, null, 2));
} catch (err) {
	if (err instanceof Prisma.PrismaClientKnownRequestError) {
		throw new Error(`(ID: ${id}): ${err.message}`);
	} else if (err instanceof Prisma.PrismaClientUnknownRequestError) {
		throw new Error(`(ID: ${id}): ${err.message}`);
	} else if (err instanceof Prisma.PrismaClientValidationError) {
		throw new Error(`(ID: ${id}): ${err.message}`);
	}
	
	throw new Error(`(ID: ${id}): unknown error`));
}
```

Even stranger, doing something like the undermentioned snippet makes `findUnique()` work, could the result be cached? (I am not familiar with Prisma's inner workings):

```typescript
try {
	await this.prisma.mode.findMany(); //Added line
	const MODE = await this.prisma.mode.findUnique({ where: { id } }); //Now returns the DB's record

	if (!MODE) {
		throw new Error(`(ID: ${id}): mode not found`));
	}

	console.log(JSON.stringify(MODE, null, 2));
} catch (err) {
	if (err instanceof Prisma.PrismaClientKnownRequestError) {
		throw new Error(`(ID: ${id}): ${err.message}`);
	} else if (err instanceof Prisma.PrismaClientUnknownRequestError) {
		throw new Error(`(ID: ${id}): ${err.message}`);
	} else if (err instanceof Prisma.PrismaClientValidationError) {
		throw new Error(`(ID: ${id}): ${err.message}`);
	}
	
	throw new Error(`(ID: ${id}): unknown error`));
}
```

Of course, all the above snippets are working in 4.9 and before.
Furthermore, I do use BigInt as the type for the `id` column, so it seems somewhat linked to Raees678's message and issue:

> This is because I can issue findUnique queries on `Int` or `BigInt` types in apollo and everything will run fine as long as Prisma's data type is not `BigInt`. If the `schama.prisma` defines it as an `Int`, the findUnique batched queries run fine, regardless if Apollo's serializer serializes it to an Int or a custom `BigInt` scalar in the response. However, if the `schema.prisma` defines it as a `BigInt` then the queries fail. So the issues must be arising from prisma's handling of the data being returned not any problem with apollo unable to serialize responses properly.

I did not find an open or closed issue on this subject, so I thought about leaving a comment here in case this message helps narrow down this problem's origin.

**Reply by kwesterfeld2** [reply]
Is this issue being actively worked?  I have hunted all over the place to solve the exact issue here -- using bigint with findUnique is broken.  If I make a resolver to use findMany it works

**Reply by edlundin** [reply]
Hi, according to #18171, it should have been fixed around the 4.12.0 release.
Now, I can't confirm this, unfortunately. Using prisma on Apple M1, I am plagued by other issues like #18510.

If at all possible, I suggest sticking to the 4.9.0 release. It is by far the more stable version, for my particular setup.

**Reply by kwesterfeld2** [reply]
I also thought 4.12 would have fixed it, and tried that and didn't have success.  It may be worth trying this again...I won't rule out my attempt could have ended up messing up dependencies but I thought I was careful here.

I didn't think about going back to 4.9.0.   That could be a very viable option.

**Comment by kwesterfeld2** [comment]
I also can confirm that if I change from using findUnique to findUniqueOrThrow, the behavior changes to the correct behavior.  Using BigInt as a model id data type just doesn't work with findUnique.

This is problematic for me because the entire model I am consuming uses BigInt exclusively for keys.  The use of findUnique is burned into nexus-prisma, which is my graphql middleware, so whenever this bit of code is hit I receive an error about non-nullable fields.

**Comment by kwesterfeld2** [comment]
This discussion was most helpful, but doesn't seem to lead to an actual software change. I created an open issue here: https://github.com/prisma/prisma/issues/19737

**Comment by baladao** [comment]
My model has an id based on an enum and suffers from the same issue. I am on prisma 5.0 tho. Using a `findMany` just before the `findUnique` makes it work, but `findUniqueOrThrow` doesn't fix the issue.

```prisma
enum ProductCategoryEnum {
  APPAREL
  ACCESSORIES
  FOOD_AND_BEVERAGE
}

model ProductCategory {
  id        ProductCategoryEnum @id
  name      String
  createdAt DateTime            @default(now()) @map("created_at")
  updatedAt DateTime            @updatedAt @map("updated_at")

  linkedTypes ProductType[]
  products    Product[]

  @@map("product_categories")
}
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
